### ISSUE TITLE
Access SlimerJS 0.10pre Webdriver from Python

### ISSUE THREAD

**Issue by RyPoints** [laurentj/slimerjs#453]
> So please, download SlimerJS 0.10pre, and test it with your Selenium installation. And open new issues if you have some problem, by explaining how to reproduce it. Thank you :)

**Anyone successfully interface the experimental SlimerJS 0.10pre with Selenium in Python?**  I'm really excited about the new webdriver feature.  If you can get me up and running and the functionality is there, I'll be happy to contribute more.

I started with this simple Python script:

```
from selenium import webdriver
+other selenium includes not used in this example

driver = webdriver.PhantomJS(executable_path="./slimerjs", service_args=['--webdriver=127.0.0.1:53598'])
#driver = webdriver.Firefox(executable_path="./slimerjs --webdriver=127.0.0.1:53598")
driver.set_window_size(1280,1024)
driver.get('https://slimerjs.org')
driver.save_screenshot('screenie.png')
```

With **driver = webdriver.PhantomJS(executable_path="./slimerjs", service_args=['--webdriver=127.0.0.1:53598'])** Firefox opens and displays 'SlimmerJS' but then errors out giving me:

```
Traceback (most recent call last):
  File "skinny.py", line 10, in <module>
    driver = webdriver.PhantomJS(executable_path="./slimerjs", service_args=['--webdriver=127.0.0.1:53598'])
  File "/usr/local/lib/python2.7/site-packages/selenium/webdriver/phantomjs/webdriver.py", line 51, in __init__
    self.service.start()
  File "/usr/local/lib/python2.7/site-packages/selenium/webdriver/phantomjs/service.py", line 83, in start
    "Can not connect to GhostDriver on port {}".format(self.port))
selenium.common.exceptions.WebDriverException: Message: Can not connect to GhostDriver on port 54658
```

The listed port it claims to not be able to connect to is different every time and doesn't match the command line port parameter.

With **driver = webdriver.Firefox(executable_path="./slimerjs --webdriver=127.0.0.1:53598")** Firefox opens, but not headlessly.  It functions just as if slimerjs wasn't there.

**Any idea how I should configure this simple Selenium script in Python?**

**Comment by laurentj** [comment]
Ok, so you probably found a bug with the port setting. I need to verify that.

**Comment by mucaho** [comment]
+1, also excited about this feature. I'd like to do WebGL testing, which isn't provided by `phantomjs`.
I'm currently investigating another option for doing webdriver tests on `webkit` - [`electron`](https://github.com/electron/electron/blob/master/docs/tutorial/using-selenium-and-webdriver.md), just to point that out for anybody interested (remember `slimerjs` is based on `gecko`).

**Comment by laurentj** [comment]
I think it is related to issue #462 : there was a bug with "--webdriver". Please try with the latest 0.10pre and reopen the issue if it doesn't work for you.

**Comment by RyPoints** [comment]
I'll try the latest .10pre soon and let you know the result.

**Comment by RyPoints** [comment]
So, I downloaded the April 27th version of .10pre.  That version didn't include the slimerjs executable script, so I copied that script over.  When I attempt to run the previously mention script I see:

```
Traceback (most recent call last):
  File "script.py", line 3, in <module>
    driver = webdriver.PhantomJS(executable_path="./slimerjs", service_args=['--webdriver=127.0.0.1:63099'])
  File "/usr/local/lib/python2.7/site-packages/selenium/webdriver/phantomjs/webdriver.py", line 51, in __init__
    self.service.start()
  File "/usr/local/lib/python2.7/site-packages/selenium/webdriver/phantomjs/service.py", line 83, in start
    "Can not connect to GhostDriver on port {}".format(self.port))
selenium.common.exceptions.WebDriverException: Message: Can not connect to GhostDriver on port 63238
```

That's the same error as before.  The error appears to indicate, for starters, that the '--webdriver=127.0.0.1:53598' parameter does not function with slimmerjs, because otherwise GhostDriver would be waiting on port 53598 and not 63238.

I'd recommend reviewing https://pypi.python.org/pypi/selenium.  Is slimerjs compatible with the existing phantomjs Python bindings?  That's the first thing to look into.

I'm happy to help in any way I can.  Can you reopen this issue?

**Comment by laurentj** [comment]
> That version didn't include the slimerjs executable script, so I copied that script over from the previous version

This is strange. The current zip contains the slimerjs script...

I tried to use your python script. The problem is that you don't use the webdriver object as expected. You don't have to indicate arguments `--webdriver` as the python driver adds it automatically. So, your script should be:

```python
from selenium import webdriver

driver = webdriver.PhantomJS(executable_path="./slimerjs")
driver.set_window_size(1280,1024)
driver.get('https://slimerjs.org')
driver.save_screenshot('screenie.png')
```

You can indicate a specific port and a file log if you want:

```python
driver = webdriver.PhantomJS(executable_path="./slimerjs", port=53645, service_log_path="slimer.log")
```

So, the current issue is solved.

**Comment by RyPoints** [comment]
Ah, okay.  I'll redownload .10pre, attempt using the revised script and let you know how things go soon.

**Comment by domderen** [comment]
Hey,

I'm trying to get the SlimerJS Selenium WebDriver working inside Docker container. It starts nicely, but unfortunately I am not able to use SlimerJS as a Selenium node.

I am currently running into two issues:
1. If I try to run the selenium runner through Selenium Grid, it fails with the error:

```
Error forwarding the new session new session request for webdriver should contain a location header or an 'application/json;charset=UTF-8' response body with the session ID.
```

2. If I try to connect to the SlimerJS WebDriver directly I just get a `Bad Request` response, and no more info.

I created a separate repository that creates the Docker setup of the WebDriver available here: https://github.com/domderen/slimerjs-docker-selenium

I'd love to help in solving those issues, and making SlimerJS an awesome headless selenium driver, but I could use some help pinpointing the issues.

Thanks!

**Comment by laurentj** [comment]
@domderen not related to this issue. please open a new issue (if not a duplicate of your issue #462), and give more details (with logs etc). And run SlimerJS *outside* docker, with debug mode, to have more info . Thank you.

**Comment by RyPoints** [comment]
I downloaded the May 2nd version of SlimerJS 1.0.  This version now re-includes the 'slimerjs' file.  I then ran your script:

```
from selenium import webdriver

driver = webdriver.PhantomJS(executable_path="./slimerjs")
driver.set_window_size(1280,1024)
driver.get('https://slimerjs.org')
driver.save_screenshot('screenie.png')

```
Firefox launches and brings up the slimerjs website in a window.  However, the screenshot is completely blank.

If this did work driver.save_screenshot should be looked at.  I'll continue testing and see if other aspects are functional.

**Comment by laurentj** [comment]
It is blank with PhantomJS too with me. It seems save_screenshot() is called before the page is loaded..

**Comment by RyPoints** [comment]
Hmmm....I thought about that, added a sleep(5) after driver.get('https://slimerjs.org') and it was still blank though.  I didn't try it with PhantomJS yet, but I will momentarily.

**Comment by RyPoints** [comment]
Ah, okay, the problem was the version I was editing in Sublime Text was still the April 27th one, but the version I was running was the May 2nd one.  After editing the correct file and adding the time.sleep(5) slimmerjs saves the screenshot.

Now time to see if this can run in AWS Lambda... ;)
